src/coreclr/vm/amd64/jitinterfaceamd64.cpp

```diff
@@ -296,6 +296,7 @@ int WriteBarrierManager::ChangeWriteBarrierTo(WriteBarrierType newWriteBarrier,
     {
         ExecutableWriterHolder<void> writeBarrierWriterHolder(GetWriteBarrierCodeLocation((void*)JIT_WriteBarrier), GetCurrentWriteBarrierSize());
         memcpy(writeBarrierWriterHolder.GetRW(), (LPVOID)GetCurrentWriteBarrierCode(), GetCurrentWriteBarrierSize());
+        stompWBCompleteActions |= SWB_ICACHE_FLUSH;
```

Comment:
It sounds like `SWB_ICACHE_FLUSH` is required on all write barrier updates now. Are there any situations where it is not required/set?

Reply:
It is needed whenever we change the code of the barrier. So for this `ChangeWriteBarrierTo`, it is needed always, because we always switch the write barrier kind. Only the `WriteBarrierManager::UpdateEphemeralBounds` and `WriteBarrierManager::UpdateWriteWatchAndCardTableLocations` have it optional, since if the bounds and barrier kind doesn't change, there is no code change.

I am actually considering making most of these changes conditional on W^X enabled. Since without W^X enabled, things work fine even with the emulation. That's because the emulator doesn't see the connection between the write location and execute location.

I just wonder - why were we flushing the instruction cache for some of the changes, but not for all of them before?

Reply:
> WriteBarrierManager::UpdateEphemeralBounds and WriteBarrierManager::UpdateWriteWatchAndCardTableLocations have it optional, since if the bounds and barrier kind doesn't change, there is no code change.

`UpdateEphemeralBounds` and `UpdateWriteWatchAndCardTableLocations ` will change immediates baked into the code. The emulator JIT has to flush the caches and rejit the code.

Reply:
> I just wonder - why were we flushing the instruction cache for some of the changes, but not for all of them before?

This is not the first time we are trying to make the updateable code work on some sort of emulator. Some of the existing hard flushes are left-overs from the earlier efforts.

Reply:
Ah, makes sense

Reply:
> UpdateEphemeralBounds and UpdateWriteWatchAndCardTableLocations will change immediates baked into the code. The emulator JIT has to flush the caches and rejit the code.

Only if the limits really change. That's not always the case.